pvlib/modelchain.py

```diff
@@ -632,30 +655,31 @@ def infer_dc_model(self):
                              'set the model with the dc_model kwarg.')
 
     def sapm(self):
-        self.dc = self.system.sapm(self.effective_irradiance,
-                                   self.cell_temperature)
+        self.results.dc = self.system.sapm(self.results.effective_irradiance,
```

Comment:
I like where this `results` attribute is going. Are we going to have a deprecation period or just call it a breaking change for 0.9? I'm fine with either. If we go with a breaking change here the maybe do the same in `PVSystem`.

Reply:
Was thinking about that. We could copy from `ModelChain.results` to the old attributes, to avoid breaking user code. Not sure how to raise a deprecation message other than to always issue a warning in `run_model`.

Reply:
I think there's a way to use the deprecation wrapper for attributes but I don't know the details.

Reply:
Looks like that capability is relatively straightforward from py3.7, use __getattr__ to intercept an attempt to access a deprecated attribute, issue warning and return the appropriate field of `results`. [PEP562](https://www.python.org/dev/peps/pep-0562/)

Reply:
That PEP is specific to adding support for *modules* to the python `getattr` *function*.  You're onto a good idea but it's not the same thing. Here I think you're wanting to define a `ModelChain.__getattr__` *method*, and within that use the python *function* to access the attribute of the results object. Something like this should be close:

```python
class ModelChain:
    def __getattr__(self, key):
        deprecated_attrs = [a list of strings]
        if key in deprecated_attrs:
            warnings.warn(...)
            return getattr(self.results, key)
        else:
            return self.__dict__[key]  # maybe wrap in try: except KeyError: raise AttributeError
```

Reply:
Is it possible to add deprecation to to attributes via `@property`? What is the advantage of `getattr` over this:

```python

class ModelChain:
    @property
    def ac(self):
        warnings.warn(...)
        return self.results.ac
```

Reply:
Good idea. I think that `@property` would work too. Maybe the main advantage of `getattr` in this situation is defining the deprecated attributes once in a list and avoiding the need to repeat the pattern for each attribute. I'm ok with either approach if they work.